Fix unified-sidebar scrolling

cpapazoglou opened the pull request:
After a lot of try & fail, I eventually:
- used vanilla JS no without passing styles through react to the sidebar component. This way we avoid rerendering the component while scrolling.
- had to manipulate the `secondary` area and not the sidebar itself as it is encapsulated over there

#### Changes proposed in this Pull Request

* Adds `...props` pass through to `Sidebar` component.
* Adds `window.addEventListener( 'scroll')` in `MySitesSidebarUnified` component so that we reposition (scroll the sidebar) like wp-admin behaves.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* load calypso with `?flags=nav-unification`
* test that the sidebar behaves as in wp-admin, especially when the window height is shorter than the sidebar
* test that my sites, reader, me work as expected
* test the mobile view (sidebar full width)

Relates #46643

mreishus commented:
I don't think the scroll handler is being cleaned up properly.

To test, apply diff:
```diff
diff --git a/client/my-sites/sidebar-unified/index.jsx b/client/my-sites/sidebar-unified/index.jsx
index 6763389a39..6296943c20 100644
--- a/client/my-sites/sidebar-unified/index.jsx
+++ b/client/my-sites/sidebar-unified/index.jsx
@@ -39,8 +39,10 @@ export const MySitesSidebarUnified = ( { path } ) => {
        const sidebarIsCollapsed = useSelector( getSidebarIsCollapsed );
 
        useEffect( () => {
+               console.log( 'Registered event listener' );
                window.addEventListener( 'scroll', () => handleScroll() );
                return () => {
+                       console.log( 'Cleared event listener' );
                        window.removeEventListener( 'scroll', () => handleScroll() );
                };
        } );
diff --git a/client/my-sites/sidebar-unified/utils.jsx b/client/my-sites/sidebar-unified/utils.jsx
index 22c82acb7b..bdf43d71f8 100644
--- a/client/my-sites/sidebar-unified/utils.jsx
+++ b/client/my-sites/sidebar-unified/utils.jsx
@@ -6,6 +6,7 @@ let pinnedSidebarBottom = false;
 let ticking = false; // Used for Scroll event throttling.
 
 export const handleScroll = () => {
+       console.log( 'handlescroll' );
        const windowHeight = window?.innerHeight;
        const secondaryElHeight = secondaryEl?.scrollHeight;
        const masterbarHeight = document.getElementById( 'header' ).getBoundingClientRect().height;
```

* Visit http://calypso.localhost:3000/home/testsite5mmr.wordpress.com?flags=nav-unification
* Open console, see registered, cleared, registered, cleared, registered
* Click "Switch Site", "All my Sites", see "Select a site" with no sidebar
![2020-10-27_17-20](https://user-images.githubusercontent.com/937354/97368380-caa6c000-1878-11eb-8e17-e9e6840068bb.png)
![2020-10-27_17-20_1](https://user-images.githubusercontent.com/937354/97368386-ce3a4700-1878-11eb-9020-4646ff104e2a.png)
![2020-10-27_17-20_2](https://user-images.githubusercontent.com/937354/97368390-d1353780-1878-11eb-93bd-8d0c37fed0c5.png)
* See 'Cleared event listener' in console
* Scroll mouse wheel
* Expect to see: No handlescroll messages  Actually see: Handlescroll messages in console

I think because A and B are different anonymous functions, it can't match them up:
```js
	useEffect( () => {
		window.addEventListener( 'scroll', () => handleScroll() ); // A
		return () => {
			window.removeEventListener( 'scroll', () => handleScroll() ); // B
		};
	} );
```

I'm about to head out, so I haven't tested it, but I think `const x = useCallback(() => handleScroll());` then using `x` in both `addEventListener` and `removeEventListener` should fix.

frontdevde commented:
From a purely functional perspective this seems to be working fine for me locally until I interact with the `My Sites` switcher which is when I start seeing the following behaviour:

![my-sites-scroll](https://user-images.githubusercontent.com/1562646/97424927-b4990e00-1911-11eb-8415-4962596dd4fb.gif)

frontdevde commented:
From a purely functional perspective this seems to be working fine for me locally until I interact with the My Sites switcher. When scrolling while the site list is showing, it starts detaching the sidebar. Sent you a screen-recording via DM.

cpapazoglou commented:
> I don't think the scroll handler is being cleaned up properly.
> 
> To test, apply diff:
> 
> ```diff
> diff --git a/client/my-sites/sidebar-unified/index.jsx b/client/my-sites/sidebar-unified/index.jsx
> index 6763389a39..6296943c20 100644
> --- a/client/my-sites/sidebar-unified/index.jsx
> +++ b/client/my-sites/sidebar-unified/index.jsx
> @@ -39,8 +39,10 @@ export const MySitesSidebarUnified = ( { path } ) => {
>         const sidebarIsCollapsed = useSelector( getSidebarIsCollapsed );
>  
>         useEffect( () => {
> +               console.log( 'Registered event listener' );
>                 window.addEventListener( 'scroll', () => handleScroll() );
>                 return () => {
> +                       console.log( 'Cleared event listener' );
>                         window.removeEventListener( 'scroll', () => handleScroll() );
>                 };
>         } );
> diff --git a/client/my-sites/sidebar-unified/utils.jsx b/client/my-sites/sidebar-unified/utils.jsx
> index 22c82acb7b..bdf43d71f8 100644
> --- a/client/my-sites/sidebar-unified/utils.jsx
> +++ b/client/my-sites/sidebar-unified/utils.jsx
> @@ -6,6 +6,7 @@ let pinnedSidebarBottom = false;
>  let ticking = false; // Used for Scroll event throttling.
>  
>  export const handleScroll = () => {
> +       console.log( 'handlescroll' );
>         const windowHeight = window?.innerHeight;
>         const secondaryElHeight = secondaryEl?.scrollHeight;
>         const masterbarHeight = document.getElementById( 'header' ).getBoundingClientRect().height;
> ```
> 
> * Visit http://calypso.localhost:3000/home/testsite5mmr.wordpress.com?flags=nav-unification
> * Open console, see registered, cleared, registered, cleared, registered
> * Click "Switch Site", "All my Sites", see "Select a site" with no sidebar
>   ![2020-10-27_17-20](https://user-images.githubusercontent.com/937354/97368380-caa6c000-1878-11eb-8e17-e9e6840068bb.png)
>   ![2020-10-27_17-20_1](https://user-images.githubusercontent.com/937354/97368386-ce3a4700-1878-11eb-9020-4646ff104e2a.png)
>   ![2020-10-27_17-20_2](https://user-images.githubusercontent.com/937354/97368390-d1353780-1878-11eb-93bd-8d0c37fed0c5.png)
> * See 'Cleared event listener' in console
> * Scroll mouse wheel
> * Expect to see: No handlescroll messages  Actually see: Handlescroll messages in console
> 
> I think because A and B are different anonymous functions, it can't match them up:
> 
> ```js
> 	useEffect( () => {
> 		window.addEventListener( 'scroll', () => handleScroll() ); // A
> 		return () => {
> 			window.removeEventListener( 'scroll', () => handleScroll() ); // B
> 		};
> 	} );
> ```
> 
> I'm about to head out, so I haven't tested it, but I think `const x = useCallback(() => handleScroll());` then using `x` in both `addEventListener` and `removeEventListener` should fix.

Great finding @mreishus , thanks! Since the callback has no dependencies, declaring
`const scrollCallback = () => handleScroll();` outside of the component and using `scrollCallback` in the event listener is enough bccd0ff

cpapazoglou commented:
> From a purely functional perspective this seems to be working fine for me locally until I interact with the My Sites switcher. When scrolling while the site list is showing, it starts detaching the sidebar. Sent you a screen-recording via DM.

Thanks @frontdevde for uncovering this! Fixed here 8a74b4c

cpapazoglou commented:
Since the changes are regarding the `secondary` and not the sidebar, I moved the code to `wp-calypso/client/layout/index.jsx` and also rebased.

This change also helps `reader` and `me` behave the same way.

obenland commented on client/layout/index.jsx:
```diff
@@ -78,6 +80,15 @@ class Layout extends Component {
 			}
 		}
+		if ( config.isEnabled( 'nav-unification' ) ) {
+			window.addEventListener( 'scroll', scrollCallback );
```
Could this be simplified to the following? 
```suggestion
			window.addEventListener( 'scroll', handleScroll );
```

cpapazoglou replied:
This was done so that we can effectively clear the event listener as discussed here https://github.com/Automattic/wp-calypso/pull/46689#issuecomment-718101323. I have already tried using the exported function `handleScroll` directly in vain. I may be doing something wrong though..

obenland commented on client/layout/utils.ts:
```diff
@@ -14,3 +14,114 @@ export function getShouldShowAppBanner( site: any ): boolean {
+let sidebarTop = 0; // Current sidebar top position.
+let pinnedSidebarTop = true;
+let pinnedSidebarBottom = false;
+let ticking = false; // Used for Scroll event throttling.
```
Is the purpose of `ticking` to indicate active scrolling?

cpapazoglou replied:
`ticking:true` indicates that we are currently animating or that a scroll happens while we are in the same animation frame. So when `ticking:true` we should not run all the calculations again. It is only used for performance reasons.

frontdevde commented:
I'm loving the direction we're going in here. This already feels way better than current Calypso. Scrolling the sidebar by scrolling the page is so much more intuitive than having to scroll inside the sidebar. 👏

Have we tested this on an iPad (in landscape and switching between orientations)?

One thing I noticed is that it's still possible to recreate a detached state when using the dev tools. We might need to factor in `resize` in this as well.

![nav-dev-tools](https://user-images.githubusercontent.com/1562646/97554577-43239300-19d7-11eb-8706-06284c957a6c.gif)

frontdevde commented:
Suggestion: Let's remove the `Fixes #46643` keyword from the description (or create a separate issue for the second part of the issue).

This PR fixes the first part of the issue, which is great! We should definitely limit the scope of this PR to that issue. That said, we'll still need to address the my site switcher overflow being visible during the transition when a site is selected. This will probably involve dynamically hiding the overflow for the duration of the transition.

cpapazoglou commented:
I have added the `resize` event which hopefully will fix any problems on orientation change. Lets merge it so I can test it on a real device easier.

frontdevde commented on client/layout/utils.ts:
```diff
@@ -21,13 +21,14 @@ let pinnedSidebarTop = true;
 
-export const handleScroll = (): void => {
+export const handleScroll = ( event: any ): void => {
+	console.log( event );
```
We probably don't want to commit logging this event?

cpapazoglou replied:
Done eb3d162, thanks!

frontdevde reviewed (approved):
After some more testing I can confirm that the resize is now handled well and the previously reported issue is gone. 

Let's remove the rouge console.log and then I'd be fine with merging it so we can a) get it in in time for demo and b) you can properly test this on the iPad. Let's keep iterating! 👍
